File: pwiz/pwiz/analysis/common/ZeroSampleFiller.cpp
//
// $Id$
//
//
// Original author: Matt Chambers <matt.chambers <a.t> vanderbilt.edu>
//
// Copyright 2008 Vanderbilt University - Nashville, TN 37232
//
// Licensed under the Apache License, Version 2.0 (the "License"); 
// you may not use this file except in compliance with the License. 
// You may obtain a copy of the License at 
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software 
// distributed under the License is distributed on an "AS IS" BASIS, 
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
// See the License for the specific language governing permissions and 
// limitations under the License.
//

#define PWIZ_SOURCE


#include "pwiz/utility/misc/Std.hpp"
#include "ZeroSampleFiller.hpp"
#include <cmath>


namespace pwiz {
namespace analysis {


void ZeroSampleFiller::fill(const vector<double>& x, const vector<double>& y,
                            vector<double>& xFilled, vector<double>& yFilled,
                            size_t zeroSampleCount)
{
    if (x.size() != y.size())
        throw runtime_error("[ZeroSampleFiller::fill()] x and y arrays must be the same size");

    // adjacent samples are expected to be within this tolerance
    const static double EPSILON = 1e-5;

    // start with the original data
    xFilled.assign(x.begin(), x.end());
    yFilled.assign(y.begin(), y.end());

    // insert flanking zeros around non-zero data points
    bool wasInData = false;
    bool nowInData = false;
    for (int i=y.size()-1; i >= 0; --i)
    {
        nowInData = yFilled[i] > 0.0;
        if (nowInData && !wasInData)
        {
            // step forward to check for missing samples
            
            // at i==0, fudge the first order delta
            double firstOrderDelta = i < 1 ? xFilled[i+1] - xFilled[i]
                                           : xFilled[i] - xFilled[i-1];

            // at i==1 or when possibly between signals, assume no second order delta
            double secondOrderDelta = i < 2 || yFilled[i-1] == 0 ? 0 :
                                      firstOrderDelta - (xFilled[i-1] - xFilled[i-2]);
            double totalDelta = 0;
            for (int j=1; j <= (int) zeroSampleCount; ++j)
            {
                totalDelta += secondOrderDelta;
                double newX = xFilled[i+j-1] + firstOrderDelta + totalDelta;
                
                bool oob = i+j >= (int) y.size();

                // sampleDelta should never be less than negative firstOrderDelta
                double sampleDelta = oob ? 0 : xFilled[i+j] - newX;
                if (sampleDelta < -firstOrderDelta)
                    break;
                    //throw std::runtime_error("[ZeroSampleFiller::fill()] miscalculated sample rate");

                // if out of bounds or newX is a valid missing sample, insert a new zero point
                if (oob || sampleDelta > firstOrderDelta)
                {
                    xFilled.insert(xFilled.begin()+(i+j), newX);
                    yFilled.insert(yFilled.begin()+(i+j), 0.0);
                }
            }
        }
        wasInData = nowInData;
    }

    wasInData = false;
    for (int i=0, end=yFilled.size(); i < end; ++i, end=yFilled.size())
    {
        nowInData = yFilled[i] > 0.0;
        if (nowInData && !wasInData)
        {
            // step backward to check for missing samples

            // at i==end-1, fudge the first order delta
            double firstOrderDelta = i == end-1 ? xFilled[i] - xFilled[i-1]
                                                : xFilled[i+1] - xFilled[i];

            // at i==end-2 or when possibly between signals, assume no second order delta
            double secondOrderDelta = i == end-2 || yFilled[i+1] == 0 ? 0 :
                                      (xFilled[i+2] - xFilled[i+1]) - firstOrderDelta;
            double totalDelta = 0;
            for (int j=1; j <= (int) zeroSampleCount; ++j)
            {
                totalDelta += secondOrderDelta;
                double newX = (xFilled[i-j+1] - firstOrderDelta) - totalDelta;

                bool oob = i-j < 0;

                // xFilled[i-j] and newX should be nearly equal if they are the same sample

                // sampleDelta should never be greater than firstOrderDelta
                double sampleDelta = oob ? 0 : xFilled[i-j] - newX;
                if (sampleDelta > firstOrderDelta)
                    break;
                    //throw std::runtime_error("[ZeroSampleFiller::fill()] miscalculated sample rate");

                // if out of bounds or newX is a valid missing sample, insert a new zero point
                if (oob || sampleDelta <= -firstOrderDelta )
                {
                    xFilled.insert(xFilled.begin()+(i-j+1), newX);
                    yFilled.insert(yFilled.begin()+(i-j+1), 0.0);
                    ++i;
                }
            }
        }
        wasInData = nowInData;
    }
}


} // namespace analysis
} // namespace pwiz

File: pwiz-sharp/pwiz/src/Analysis/PeakPicking/ZeroSampleFiller.cs
namespace Pwiz.Analysis.PeakPicking;

/// <summary>
/// Fills missing zero samples around signal profiles so peak-picking algorithms see a well-formed
/// flanking context. Port of <c>pwiz::analysis::ZeroSampleFiller</c>.
/// </summary>
/// <remarks>
/// The pre-condition: the sample rate may change but must change gradually, and there is at
/// least one zero sample on each side of every non-zero run (so the first-order delta can be
/// read off existing samples without extrapolation).
/// </remarks>
public static class ZeroSampleFiller
{
    /// <summary>
    /// Populates <paramref name="xFilled"/> / <paramref name="yFilled"/> with the input samples
    /// plus up to <paramref name="zeroSampleCount"/> zero samples on each flank of every
    /// non-zero run. Existing samples are preserved.
    /// </summary>
    public static void Fill(
        IReadOnlyList<double> x,
        IReadOnlyList<double> y,
        List<double> xFilled,
        List<double> yFilled,
        int zeroSampleCount)
    {
        ArgumentNullException.ThrowIfNull(x);
        ArgumentNullException.ThrowIfNull(y);
        ArgumentNullException.ThrowIfNull(xFilled);
        ArgumentNullException.ThrowIfNull(yFilled);
        if (x.Count != y.Count)
            throw new ArgumentException("x and y arrays must be the same size.");

        xFilled.Clear();
        yFilled.Clear();
        xFilled.AddRange(x);
        yFilled.AddRange(y);

        // Scan backward, inserting flanking zeros to the RIGHT of the last non-zero sample of
        // each run (where the run transitions from data → zero as we walk from right to left).
        bool wasInData = false;
        for (int i = yFilled.Count - 1; i >= 0; i--)
        {
            bool nowInData = yFilled[i] > 0.0;
            if (nowInData && !wasInData)
            {
                // at i == 0, fudge the first-order delta
                double firstOrderDelta = i < 1 ? xFilled[i + 1] - xFilled[i]
                                                : xFilled[i] - xFilled[i - 1];
                double secondOrderDelta = i < 2 || yFilled[i - 1] == 0
                    ? 0
                    : firstOrderDelta - (xFilled[i - 1] - xFilled[i - 2]);
                double totalDelta = 0;
                for (int j = 1; j <= zeroSampleCount; j++)
                {
                    totalDelta += secondOrderDelta;
                    double newX = xFilled[i + j - 1] + firstOrderDelta + totalDelta;
                    bool oob = i + j >= yFilled.Count;
                    double sampleDelta = oob ? 0 : xFilled[i + j] - newX;
                    if (sampleDelta < -firstOrderDelta) break;
                    if (oob || sampleDelta > firstOrderDelta)
                    {
                        xFilled.Insert(i + j, newX);
                        yFilled.Insert(i + j, 0.0);
                    }
                }
            }
            wasInData = nowInData;
        }

        // Scan forward, inserting flanking zeros to the LEFT of the first non-zero sample of
        // each run (transition from zero → data).
        wasInData = false;
        for (int i = 0; i < yFilled.Count; i++)
        {
            bool nowInData = yFilled[i] > 0.0;
            if (nowInData && !wasInData)
            {
                int end = yFilled.Count;
                double firstOrderDelta = i == end - 1
                    ? xFilled[i] - xFilled[i - 1]
                    : xFilled[i + 1] - xFilled[i];
                double secondOrderDelta = i == end - 2 || yFilled[i + 1] == 0
                    ? 0
                    : (xFilled[i + 2] - xFilled[i + 1]) - firstOrderDelta;
                double totalDelta = 0;
                for (int j = 1; j <= zeroSampleCount; j++)
                {
                    totalDelta += secondOrderDelta;
                    double newX = (xFilled[i - j + 1] - firstOrderDelta) - totalDelta;
                    bool oob = i - j < 0;
                    double sampleDelta = oob ? 0 : xFilled[i - j] - newX;
                    if (sampleDelta > firstOrderDelta) break;
                    if (oob || sampleDelta <= -firstOrderDelta)
                    {
                        xFilled.Insert(i - j + 1, newX);
                        yFilled.Insert(i - j + 1, 0.0);
                        i++;
                    }
                }
            }
            wasInData = nowInData;
        }
    }
}
